File: apps/frontend/lib/core/errors/idempotency_error_codes.dart
abstract final class IdempotencyErrorCodes {
  static const keyReused = 'IDEMPOTENCY_KEY_REUSED';
  static const inProgress = 'IDEMPOTENCY_IN_PROGRESS';
  static const previouslyFailed = 'IDEMPOTENCY_PREVIOUSLY_FAILED';

  static const byName = <String, String>{
    'keyReused': keyReused,
    'inProgress': inProgress,
    'previouslyFailed': previouslyFailed,
  };

  static const values = <String>{keyReused, inProgress, previouslyFailed};

  static bool contains(String? value) => values.contains(value);
}

File: apps/backend/src/contracts/errors/idempotency-error-codes.ts
export const IDEMPOTENCY_ERROR_CODES = {
  keyReused: 'IDEMPOTENCY_KEY_REUSED',
  inProgress: 'IDEMPOTENCY_IN_PROGRESS',
  previouslyFailed: 'IDEMPOTENCY_PREVIOUSLY_FAILED',
} as const;

export type IdempotencyErrorCode =
  (typeof IDEMPOTENCY_ERROR_CODES)[keyof typeof IDEMPOTENCY_ERROR_CODES];

const IDEMPOTENCY_ERROR_CODE_VALUES = new Set<string>(
  Object.values(IDEMPOTENCY_ERROR_CODES),
);

export const isIdempotencyErrorCode = (
  value: string | null | undefined,
): value is IdempotencyErrorCode =>
  value != null && IDEMPOTENCY_ERROR_CODE_VALUES.has(value);
